iris-GmbH/meta-iris-base
# Feature/migl/rdphoen 1153 env writeable list

With this feature, the u-boot env's to be imported by the EMMC are limited to a few.

```
#define CONFIG_ENV_FLAGS_LIST_STATIC \
       "upgrade_available:dw," \
       "bootcount:dw," \
       "ustate:dw," \
       "firmware:dw"
```

First the default env's are loaded from the u-boot (nowhere) and then the env's from the EMMC.
The special feature is that only the env's from the writelist are overwritten, all other env's from the EMMC are ignored when importing.


Tested with irma6-maintenance-uuu on the imx8mp-evk and imx8mp-irma6r2!

**Test**
```
1. Check bootlog Part 1:

Loading Environment from nowhere... OK                                                                                                                   
Loading Environment from MMC... *** Warning - bad CRC, using default environment  <-- ok, there are no env on emmc on first boot

2. Check printenv
u-boot=> printenv
firmware=0
upgrade_available=0
ustate=0
ver=U-Boot 2020.04-00002-gc4497e2c54 (Jun 14 2022 - 21:04:58 +0200)  <--- NEW !!!

3. Test alternating saveenv works
u-boot=> saveenv                                                                                                                                         
Saving Environment to MMC... Writing to MMC(2)... OK                                                                                                     
u-boot=> saveenv                                                                                                                                         
Saving Environment to MMC... Writing to redundant MMC(2)... OK                                                                                           
u-boot=> saveenv                                                                                                                                         
Saving Environment to MMC... Writing to MMC(2)... OK                                                                                                     
u-boot=> saveenv                                                                                                                                         
Saving Environment to MMC... Writing to redundant MMC(2)... OK

4. Test ENV_WRITEABLE_LIST works
u-boot=> setenv ustate 1
u-boot=> setenv my_var test_str
u-boot=> setenv firmware 3
u-boot=> saveenv
Saving Environment to MMC... Writing to MMC(2)... OK

u-boot=> printenv                                                                                                                                                                                                                                                                     
firmware=3                                                                                                                                                                                                                                                              
my_var=test_str                                                                                                                                                                                                                                                                          
ustate=1                                                                                                                                                 

u-boot=> reset 
u-boot=> printenv                                                                                                                                       
firmware=3                                                                                                                                               
[..]    <-- "my_var" is not imported, because it is not part of the ENV_WRITEABLE_LIST
ustate=1

5. Retry 4. with redunand env slot
-> "Saving Environment to MMC... Writing to redundant MMC(2)... OK"

6. Bootlog überprüfen Teil 2:

Loading Environment from nowhere... OK
Loading Environment from MMC... OK      <-- ok, after saveenv is called there are valid env's on EMMC

7. Test interaction with Linux (fw_printenv/fw_setenv):
root@imx8mpevk:~# fw_printenv
firmware=3
my_var=test_str  <--!!! Yes, my_var still exists, when reading with fw_printenv. However, it is not imported in u-boot.
upgrade_available=0
ustate=1

root@imx8mpevk:~# fw_setenv firmware 4                                                                                                                                                                   
root@imx8mpevk:~# fw_setenv ustate 2

root@imx8mpevk:~# fw_printenv
firmware=4
my_var=test_str
ustate=2

root@imx8mpevk:~#  reboot

u-boot=> printenv
firmware=4
upgrade_available=0
ustate=2

9. Use default env
u-boot=> env default -a
u-boot=> saveenv
u-boot=> reset
u-boot=> printenv
firmware=0
upgrade_available=0
ustate=0

```
